m3/src/main/java/com/uber/m3/tally/m3/M3Reporter.java

```diff
@@ -0,0 +1,781 @@
+
+        BucketPair[] bucketPairs = BucketPairImpl.bucketPairs(buckets);
+
+        for (int i = 0; i < bucketPairs.length; i++) {
+            Map<String, String> valueTags = new HashMap<>(tags);
+            Map<String, String> durationTags = new HashMap<>(tags);
+
+            String idTagValue = String.format(bucketIdFmt, i);
+
+            valueTags.put(bucketIdTagName, idTagValue);
+            valueTags.put(bucketTagName, String.format("%s-%s",
+                valueBucketString(bucketPairs[i].lowerBoundValue()),
+                valueBucketString(bucketPairs[i].upperBoundValue())));
+
+            cachedValueBuckets.add(new CachedHistogramBucketImpl(
+                bucketPairs[i].upperBoundValue(),
+                bucketPairs[i].upperBoundDuration(),
+                (CachedMetric) allocateCounter(name, valueTags)
+            ));
+
+            durationTags.put(bucketIdTagName, idTagValue);
+            durationTags.put(bucketTagName, String.format("%s-%s",
+                durationBucketString(bucketPairs[i].lowerBoundDuration()),
+                durationBucketString(bucketPairs[i].upperBoundDuration())));
+
+            cachedDurationBuckets.add(new CachedHistogramBucketImpl(
+                bucketPairs[i].upperBoundValue(),
+                bucketPairs[i].upperBoundDuration(),
+                (CachedMetric) allocateCounter(name, durationTags)
+            ));
+        }
+
+        return new CachedHistogramImpl(
+            name,
+            tags,
+            buckets,
+            cachedValueBuckets,
+            cachedDurationBuckets
+        );
+    }
+
+    @Override
+    public Capabilities capabilities() {
+        return CapableOf.REPORTING_TAGGING;
+    }
+
+    @Override
+    public void flush() {
+        try {
+            // A null metric signifies to processors that a flush was called,
+            // which causes the metric queue to be cleared
+            metricQueue.put(resourcePool.getSizedMetric());
+        } catch (InterruptedException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private List<Metric> flush(List<Metric> metrics) throws TException {
+        TException exception = null;
+
+        synchronized (metricBatch) {
+            metricBatch.setMetrics(metrics);
+
+            try {
+                client.emitMetricBatch(metricBatch);
+            } catch (TException e) {
+                // Store exception to throw later after clean up.
+                // Don't want to put everything in a finally block here,
+                // which will hold the lock unnecessarily long
+                exception = e;
+            }
+
+            metricBatch.setMetrics(null);
+        }
+
+        resourcePool.releaseShallowMetrics(metrics);
+
+        metrics.clear();
+
+        if (exception != null) {
+            throw exception;
+        }
+
+        return metrics;
+    }
+
+    @Override
+    public void close() {
```

Comment:
IIRC the semantics were ~ you have to ensure the metric queue is flushed prior to closing, yeah?